### ISSUE TITLE
[💡SUG] Top-N Evaluations on Labled data (label created by rating threshold)

### ISSUE THREAD

**Issue by irecsys** [RUCAIBox/RecBole#1086]
Can you provide a way to convert Interaction back to Dataframe?

I do see that you have a function which transforms dataframe to interaction

**Comment by irecsys** [comment]
Actually I am looking at your Interaction.py

While testing, it should be guaranteed that all interaction records of one single
        user will not appear in different Interaction object, and records of the same user
        should be continuous. Meanwhile, the positive cases of one user always need to occur
        **earlier** than this user's negative cases.

May I know where is the piece of the coding that you "guaranteed that all interaction records of one single
        user will not appear in different Interaction object, and records of the same user
        should be continuous."

**Comment by irecsys** [comment]
The case here is, I am running context-aware. In the testing phrase, the eval-data should send interaction data which have <user, context> info. But now, the fullsortDataloader delviered interaction to me in which there are only user IDs. So, I need to figure out where to add context information. It could be fullsortDataloader? or the Interaction object

**Comment by Sherry-XLL** [comment]
@irecsys If you want to utilize the user's context information, you can load it in `user_feat` and index by user ID.

**Comment by irecsys** [comment]
> @irecsys If you want to utilize the user's context information, you can load it in `user_feat` and index by user ID.

not user information, but attributes of interactions, such as time, location, etc. These are additional information or columns saved in data.inter file.

I took a look at 'eval_data', there are such context variables loaded. But when it comes to the interaction and batch_data, there are only user ids

**Comment by Sherry-XLL** [comment]
@irecsys I run WideDeep in the context-aware models on RecBole. The interaction displayed in the evaluation phase contains context information as shown below, so I'm not sure why there are only user ids in your description.

```python
def predict(self, interaction):
    print(interaction)
    exit()
    return self.forward(interaction)
```

```
python run_recbole.py --model=WideDeep --config_files=yaml/ml-1m_context.yaml
Trainable parameters: 32160
Train     0: 100%|█████████████████████████| 20/20 [00:02<00:00,  9.36it/s, GPU RAM: 0.01 G/11.91 G]
16 Dec 23:20    INFO  epoch 0 training [time: 2.16s, train loss: 13.7366]
Evaluate   :   0%|                                                            | 0/3 [00:00<?, ?it/s]The batch_size of interaction: 4096
    user_id, torch.Size([4096]), cuda, torch.int64
    item_id, torch.Size([4096]), cuda, torch.int64
    label, torch.Size([4096]), cuda, torch.float32
    age, torch.Size([4096]), cuda, torch.int64
    gender, torch.Size([4096]), cuda, torch.int64
    occupation, torch.Size([4096]), cuda, torch.int64


Evaluate   :   0%|                                                            | 0/3 [00:00<?, ?it/s]
```

**Comment by irecsys** [comment]
> @irecsys If you want to utilize the user's context information, you can load it in `user_feat` and index by user ID.

The "context-aware" in the RecBole library is a little bit misleading in comparison with the notion in academia. In the library, both user demographic information (e.g., age, gender, nationality) and item features (e.g., movie genre, actors, etc) are considered as "context".

In academic, context variabels refer to the attributes of activities or interactions (e.g., time, location), not the attributes from users or items.

Therefore, the top-N recommendation in context-aware recsys will produce a list of items to each <user, contexts> pair. In traditional recsys, a list of recommendations will be delivered to a single user.

In the current development of fullSortDataloader, the batch_data only has a list of user ids being sent to the model for predictions. Now, I would like to send <user, context> pairs to a mdoel to get predictions

**Comment by irecsys** [comment]
My question is similar to the one in this post https://github.com/RUCAIBox/RecBole/issues/891

**Comment by Sherry-XLL** [comment]
@irecsys There is something wrong with your configuration. The evaluation mode used in context-based recommendation is `labeled` rather than `full`, and context information in `.inter` such as `timestamp` will be loaded into the interaction.

```
Train     0: 100%|█████████████████████████| 20/20 [00:00<00:00, 22.80it/s, GPU RAM: 0.03 G/11.91 G]
16 Dec 23:37    INFO  epoch 0 training [time: 0.88s, train loss: 15.3897]
Evaluate   :   0%|                                                            | 0/3 [00:00<?, ?it/s]The batch_size of interaction: 4096
    user_id, torch.Size([4096]), cuda, torch.int64
    item_id, torch.Size([4096]), cuda, torch.int64
    timestamp, torch.Size([4096]), cuda, torch.float32
    label, torch.Size([4096]), cuda, torch.float32
    age, torch.Size([4096]), cuda, torch.int64
    gender, torch.Size([4096]), cuda, torch.int64
    occupation, torch.Size([4096]), cuda, torch.int64
    class, torch.Size([4096, 6]), cuda, torch.int64
```

**Comment by irecsys** [comment]
> @irecsys There is something wrong with your configuration. The evaluation mode used in context-based recommendation is `labeled` rather than `full`, and context information in `.inter` such as `timestamp` will be loaded into the interaction.
> 
> ```
> Train     0: 100%|█████████████████████████| 20/20 [00:00<00:00, 22.80it/s, GPU RAM: 0.03 G/11.91 G]
> 16 Dec 23:37    INFO  epoch 0 training [time: 0.88s, train loss: 15.3897]
> Evaluate   :   0%|                                                            | 0/3 [00:00<?, ?it/s]The batch_size of interaction: 4096
>     user_id, torch.Size([4096]), cuda, torch.int64
>     item_id, torch.Size([4096]), cuda, torch.int64
>     timestamp, torch.Size([4096]), cuda, torch.float32
>     label, torch.Size([4096]), cuda, torch.float32
>     age, torch.Size([4096]), cuda, torch.int64
>     gender, torch.Size([4096]), cuda, torch.int64
>     occupation, torch.Size([4096]), cuda, torch.int64
>     class, torch.Size([4096, 6]), cuda, torch.int64
> ```

Thanks. Swtiched to "mode: labeled", I am able to see the context info loaded into interaction and batch data.

interaction, row_idx, positive_u, positive_i = batched_data

However, positive_u, positive_i is "<class 'NoneType'>"

May you help me point out what's wrong or where to modify

**Comment by Sherry-XLL** [comment]
@irecsys When `mode: labeled`, `positive_u` and `positive_i` are useless so they are both `NoneType`.

**Comment by irecsys** [comment]
> @irecsys When `mode: labeled`, `positive_u` and `positive_i` are useless so they are both `NoneType`.

1). I guess I am confused with full and labeled

full = full ranking with all item candidates
labeled = get positive and negative labels from data

I am supposed to evaluate all item candidates, provide a list of topN, and compare it with the relevant data (they are labeled in this case). So, I should use "labeled" mode here? what's the difference between full and labeled

2). positive_u and positive_i are NoneType, and it will results in the following error in function _neg_sample_batch_eval

        if self.config['eval_type'] == EvaluatorType.VALUE:
            return interaction, origin_scores, positive_u, positive_i
        elif self.config['eval_type'] == EvaluatorType.RANKING:
            col_idx = interaction[self.config['ITEM_ID_FIELD']]
            batch_user_num = positive_u[-1] + 1 ============================> err here
            scores = torch.full((batch_user_num, self.tot_item_num), -np.inf, device=self.device)
            scores[row_idx, col_idx] = origin_scores
            return interaction, scores, positive_u, positive_i

**Comment by irecsys** [comment]
I tried it on the MovieLens data, the mode: labeled with threshold setting will pop similar errors above

field_separator: "\t"
seq_separator: " "
show_progress: False
save_dataset: False
save_dataloaders: False

dataset: ml-100k
model: NeuMF
USER_ID_FIELD: user_id
ITEM_ID_FIELD: item_id
RATING_FIELD: rating
NEG_PREFIX: neg_
LABEL_FIELD: label
threshold:
    rating: 3
load_col:
    inter: [user_id, item_id, rating]

# training and evaluation
epochs: 500
train_batch_size: 4096
eval_batch_size: 409600000

neg_sampling: ~

eval_args:
  split: {'RS': [0.8, 0.1, 0.1]}
  group_by: ~
  mode: full
  order: RO
valid_metric: Recall
metrics: ['Recall']
topk: [10]

# model
embedding_size: 64

**Comment by Sherry-XLL** [comment]
@irecsys Hello, you need to clarify the form of the recommendation task first. If you need to use context information in RecBole's context-aware models, you should use [value-based methods](https://recbole.io/docs/user_guide/config/evaluation_settings.html) such as `AUC` for evaluation, and this is the CTR (Click-Through-Rate) task. CTR task is to determine whether a user will click on an item given some context information, the relevant configurations are as follows:

```
# dataset config
field_separator: "\t"
seq_separator: " "
USER_ID_FIELD: user_id
ITEM_ID_FIELD: item_id
RATING_FIELD: rating
NEG_PREFIX: neg_
LABEL_FIELD: label
threshold:
    rating: 4
load_col:
    inter: [user_id, item_id, rating, timestamp]
    user: [user_id, age, gender, occupation]
    item: [item_id, class]

# training and evaluation
epochs: 500
train_batch_size: 4096
eval_args:
  split: {'RS':[0.8, 0.1, 0.1]}
  group_by: ~
  mode: labeled
  order: RO
valid_metric: AUC
metrics: ['AUC', 'LogLoss']
```

If you want to utilize context information for topk recommendation, that is, recommend a list of items to users. In this case, RecBole does not support adding interactive information, and you can only assist recommendation through user or item attributes.

**Comment by irecsys** [comment]
OK. Got it. I guess we need to implement out own topN evaluations in the library. Thanks for your notes

**Comment by irecsys** [comment]
I am starting to rewrite parts of the coding. But I do need your help to clarify something in your coding

def _set_user_property(self, uid, used_item, positive_item):
        """For each user id (uid), set three attributes:
            positive items, items number, history items (used, not in positive list)
        """
        if uid is None:
            return
        history_item = used_item - positive_item  # remove positive item from used id
        self.uid2positive_item[uid] = torch.tensor(list(positive_item), dtype=torch.int64)
        self.uid2items_num[uid] = len(positive_item)
        self.uid2history_item[uid] = torch.tensor(list(history_item), dtype=torch.int64)

1). what's the difference among positve_item, used_item, history_item?

2). what are the values should be saved in positive_u, positive_i?

**Comment by Sherry-XLL** [comment]
@irecsys 1). `used_item` refers to all items that users have interacted in the training and evaluation set. `positve_item` refers to all items that users have interacted in the evaluation set. `history_item` refers to all items that users have interacted in the training set.

2). `positive_u` and `positive_i` are used to calculate evaluation scores. `positive_u(Torch.Tensor)` should be the row index of positive items for each user in the evaluation set, and `positive_i(Torch.Tensor)` should be the positive item id for each user.

https://github.com/RUCAIBox/RecBole/blob/5179244330bbf889398fb95f3d3f84fd6eb61b21/recbole/evaluator/collector.py#L129-L138

**Comment by irecsys** [comment]
> @irecsys 1). `used_item` refers to all items that users have interacted in the training and evaluation set. `positve_item` refers to all items that users have interacted in the evaluation set. `history_item` refers to all items that users have interacted in the training set.
> 
> 2). `positive_u` and `positive_i` are used to calculate evaluation scores. `positive_u(Torch.Tensor)` should be the row index of positive items for each user in the evaluation set, and `positive_i(Torch.Tensor)` should be the positive item id for each user.
> 
> https://github.com/RUCAIBox/RecBole/blob/5179244330bbf889398fb95f3d3f84fd6eb61b21/recbole/evaluator/collector.py#L129-L138

great! thanks

So batchdata is acquired from training set for training purpose. The evaluation will be performed on the validation or test data. However, there are also batch data from EvalDataloader. I beleive these batch data will be used to track the running progress. The evaluations will be performed on the whole validation or test data, right? for example, the list of relevant items for each user is extracted from the whole validation data, not the batched data.

**Comment by irecsys** [comment]
Thanks for your updates. I have successfully made it!!!

Only one drawback. In the _full_sort_batch_eval() function below, it will make prediction on all items in the data set. In recommender systems, we usually recommend items which were not in training set. (we allow recommending consumed items in special cases).

Now, I have produced a list of item candidates for each user, and save them in a dict, such as dict_user_candidates, where key is userID, value is a list of candidate items.

How can I change the coding below to make sure it will predict the score for the candidate items (for each user) only?

    def _full_sort_batch_eval(self, batched_data):
        interaction, history_index, positive_u, positive_i = batched_data
        try:
            # Note: interaction without item ids
            scores = self.model.full_sort_predict(interaction.to(self.device))
        except NotImplementedError:
            inter_len = len(interaction)
            new_inter = interaction.to(self.device).repeat_interleave(self.tot_item_num) ========> all items
            batch_size = len(new_inter)
            new_inter.update(self.item_tensor.repeat(inter_len))
            if batch_size <= self.test_batch_size:
                scores = self.model.predict(new_inter)
            else:
                scores = self._spilt_predict(new_inter, batch_size)

        scores = scores.view(-1, self.tot_item_num)
        scores[:, 0] = -np.inf
        if history_index is not None:
            scores[history_index] = -np.inf
        return interaction, scores, positive_u, positive_i

**Comment by Sherry-XLL** [comment]
@irecsys RecBole currently does not support the recommendation of cold start items, and it is not easy to evaluate the recommendation quality of items that do not appear in the data set.

**Comment by irecsys** [comment]
> @irecsys RecBole currently does not support the recommendation of cold start items, and it is not easy to evaluate the recommendation quality of items that do not appear in the data set.

we do not include cold-start items. Just a list of the items in the training set, but not consumed by each user before. In other words, the items consumed by a user in the training set will not be recommended to the user again

**Comment by Sherry-XLL** [comment]
@irecsys RecBole supports two recommendation scenarios, repeatable recommendation and non-repeatable recommendation. 

`repeatable (bool)`: Whether to evaluate the result with a repeatable recommendation scene. Note that it is disabled for sequential models as the recommendation is already repeatable. For other models, defaults to `False`. 

By default, the items that users interact with in the training set will not be recommended in the test set.

**Comment by irecsys** [comment]
> @irecsys RecBole supports two recommendation scenarios, repeatable recommendation and non-repeatable recommendation.
> 
> `repeatable (bool)`: Whether to evaluate the result with a repeatable recommendation scene. Note that it is disabled for sequential models as the recommendation is already repeatable. For other models, defaults to `False`.
> 
> By default, the items that users interact with in the training set will not be recommended in the test set.

great, thanks

**Comment by irecsys** [comment]
Hello, me again. In calculating the topK metrics, such as precision and recall, we should find a list of relevant items in the test set, and calculate a match between the recommended items and the relevant items.

The relevant items have been converted to binary labels. That's easy. 

In terms of the matched relevant items in the recommendation list, it seems that you build a matrix, and use binary value or True/False to indicate whether it is relevant. I want to know which piece of the coding (and python file) I can refer to, so that I know how did you find a match between recommended items and the actual relevant items in the test/validation set.

I revised some coding. In my outputs, sometimes recall is > 1. 
Recall = number of hits / number of actual positive items.  I guess the way of calculating the number of hits has some problems. I need to investigate the piece of the coding where the hits were calculated.

Note that, this is not a bug in Recbole. It is a bug in my coding, since my application is different from the common recommendation scenario

**Comment by irecsys** [comment]
hello, need your help again.

In my coding, I got recall > 1 sometimes. This is not a bug in Recbole, since I changed some coding

Recall = number of matched relevant items in topK / number of actual relevant items

relevant items were collected from positive_u, positive_i

So, there must be something wrong in the matching beween recommendation list and the list of positive items

May I know where did you calculate the match?

**Comment by irecsys** [comment]
In your system:

positve_item refers to all items that users have interacted in the evaluation set. 
history_item refers to all items that users have interacted in the training set. (used but not in the positive list)

positive_u and positive_i are used to calculate evaluation scores. 
positive_u(Torch.Tensor) should be the row index of positive items for each user in the evaluation set
positive_i(Torch.Tensor) should be the positive item id for each user.

In my system, i used ratings, want to confirm with you whether the following logic is correct

I set a rating threshold as 3, items with rating > 3 are positive items

positve_item refers to all items that users have rating>3 in the evaluation set. 
history_item refers to all items that users have rated in the training set. (used but not in the positive list)

positive_u and positive_i are used to calculate evaluation scores. 
positive_u(Torch.Tensor) should be the row index of positive items for each user in the evaluation set
positive_i(Torch.Tensor) should be the positive item id for each user.

positive_u = the row index? in the tensor?

I did not change any coding for positive_u, positive_i. 
If I understand correctly, they can be automatically produced, if the positve_item and history_item are prepared correctly.

**Comment by Sherry-XLL** [comment]
@irecsys From your description, I can't know why there is a problem of recall greater than 1. Maybe you can try to debug based on the relevant code.

https://github.com/RUCAIBox/RecBole/blob/5179244330bbf889398fb95f3d3f84fd6eb61b21/recbole/evaluator/collector.py#L145-L153

**Comment by irecsys** [comment]
Yes. I am debug it. Can u clarify my doubts in the previous two comments?

**Comment by irecsys** [comment]
> pos_idx

I found the reason why

Take the pos_idx for example, 

Assume I am going to recommend top-10 items to 100 users. The size of pos_idx is always 100x11

In calculating topK metrics, such as precision or recall, the 11th column is always TRUE

I did not change these coding in Recbole. Is this a bug? why the size is always number of users x (N + 1)?

**Comment by irecsys** [comment]
> @irecsys From your description, I can't know why there is a problem of recall greater than 1. Maybe you can try to debug based on the relevant code.
> 
> https://github.com/RUCAIBox/RecBole/blob/5179244330bbf889398fb95f3d3f84fd6eb61b21/recbole/evaluator/collector.py#L145-L153

Finally, I found the reason why.. That's a super big in the library. See my bug report at https://github.com/RUCAIBox/RecBole/issues/1107
